cps/src/main/java/org/jenkinsci/plugins/workflow/cps/FlowHead.java

```diff
@@ -119,7 +120,9 @@ void setNewHead(FlowNode v) {
             } else {
                 // in recovering from error and such situation, we sometimes need to grow the graph
                 // without running the program.
-                execution.notifyListeners(v);
+                // TODO can CpsThreadGroup.notifyNewHead be used instead?
+                execution.notifyListeners(Collections.singletonList(v), true);
+                execution.notifyListeners(Collections.singletonList(v), false);
```

Comment:
Maybe add a comment here explaining what this is about coz it seems odd. Notifying the synchronous and "non-synchronous" listeners.

Something really does smell strange about all of this. I'm sure it's just that I don't fully understand the issue you're trying to work around.